build_tools_check.js

```diff
@@ -0,0 +1,44 @@
+const util = require('util');
+const exec = util.promisify(require('child_process').exec);
+
+let OSName = '';
+if (process.platform === 'win32')   OSName = 'Windows';
+if (process.platform === 'darwin')  OSName = 'MacOS';
+if (process.platform === 'linux')   OSName = 'Linux';
+
+const verifyBuildToolInstallation = async (tool, checkCommand, installCommand = null) => {
+	try {
+		const { stdout } = await exec(checkCommand);
+		console.log(`${tool} is Installed: ${stdout}`);
+	} catch (error) {
+		installCommand ? installBuildTool(tool, installCommand) : console.warn(`WARNING: Please ensure that ${tool} is installed on your system`);
+	}
+};
+
+const installBuildTool = async (tool, command) =>  {
+	const message = `Installing ${tool}....`;
+	console.log(message);
+	try {
+		const { stdout } = await exec(command);
+		console.log(stdout);
+		console.log(`${tool} Installed Successfully`);
+	} catch (error) {
+		console.log(error);
+		console.log(`Something went wrong, Please ensure that ${tool} is installed in your system before proceeding`);
+	}
+};
+
+switch (OSName) {
+case 'Windows':
+	verifyBuildToolInstallation('Windows Build Tools', 'npm ls -g windows-build-tools', 'npm install --global windows-build-tools');
+	break;
+case 'MacOS':
+	verifyBuildToolInstallation('CocoaPods', 'pod --version', 'sudo gem install cocoapods');
+	verifyBuildToolInstallation('Rsync', 'rsync -version');
+	break;
+case 'Linux':
+	verifyBuildToolInstallation('Rsync', 'rsync -version', 'apt-get install rsync');
```

Comment:
This won't work on rpm-based systems. You'll have to find a better way. There are a bunch of other tools as well: `yum`, `dnf`, `emerge`, `pacman`, ....

Reply:
Yes i did some research and found this out, but following Laurent's statement in #2631 :
> Building the app requires a few external tools such as CocoaPods on macOS, rsync on Linux and macOS and windows-build-tools on Windows.

I thought he meant that we should just carry out the check for the 3 main OS: Linux, macOS and Windows.

That said, the `verifyBuildToolInstallation` function accepts 3 parameters which are:
-  `name` - The name of the dependency/build tool
- `checkCommand` - the command that will be used to check/verify if the tool is installed
- `installCommand` - the command that will be ran if the tool is not installed. This parameter is optional,  it has a default value of `null` so if `installCommand` is not provided then a message is printed to the command line to report to the user that he/she has to ensure that the tool is installed before he proceeds. which syncs with Laurent's suggestion here:
> or perhaps there's a way to auto-detect these dependencies when building the app and **report it to the user**.

So basically, We can simply omit the `installCommand` for linux systems and leave the function to simply check and report to the user that the necessary build tools are missing and he/she should install it OR we could simply add the `installCommand` and it tries to run the command for installing rsync with `sudo apt install rsync` but if it fails, it reports to the user to install it on his own before proceeding. This logic can be found here :
```
  try  {
		const { stdout } = await exec(command);
		console.log(stdout);
		console.log(`${tool} Installed Successfully`);
	} catch (error) {
		console.error(error);
		console.log(`Something went wrong, Please ensure that ${tool} is installed in your system before proceeding`);
	}
```


This explanation is just to clarify my understanding of the problem based on Laurent's statements in the issue, so if my understanding is wrong please let me know how exactly i am meant to go about this so that i can retrace my steps and perfect this solution. Thanks

Reply:
Don't try to run the install commands from the script, it would be way too complex to get this working. Just display your command as a guideline. eg. "This development tool is not installed: "rsync". Please install it using your package manager. For example `apt install rsync`"

Reply:
Awesome, I thought as much. Thanks Laurent. 

I’ll make the changes as soon as possible.

Reply:
I have made the requested change

Reply:
Please check the other inline comment.